segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,385 @@
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            return this.getStreamSegmentInfo(streamSegmentName, timeout).get();
+        } catch (Exception e) {
+            log.info("Exception {} while creating a segment {}", e, streamSegmentName);
+            if (e instanceof FileAlreadyExistsException) {
+                throw new CompletionException(new StreamSegmentExistsException(streamSegmentName, e));
+            } else {
+                throw new CompletionException(e);
+            }
+        }
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout) {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (handle.isReadOnly()) {
+            log.info("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            throw new CompletionException(new IllegalArgumentException());
+        }
+
+        if (!Files.exists(path)) {
+            throw new CompletionException(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+        }
+
+        try {
+            if (!isWritableFile(path)) {
+                throw new CompletionException(new StreamSegmentSealedException(handle.getSegmentName()));
+            }
+
+            long fileSize = path.toFile().length();
+            if (fileSize < offset) {
+                throw new CompletionException(new BadOffsetException(handle.getSegmentName(), fileSize, offset));
+            } else {
+                try (FileChannel channel = FileChannel.open(path,
+                        StandardOpenOption.WRITE); ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    channel.force(true);
+                }
+                return null;
+            }
+        } catch (IOException exc) {
+            log.info("Write to segment {} at offset {} failed with exception {} ", handle.getSegmentName(), offset,
+                    exc.getMessage());
+            if (exc instanceof AccessDeniedException) {
+                throw new CompletionException(new IllegalStateException(handle.getSegmentName()));
+            } else if (exc instanceof ClosedChannelException) {
+                throw new CompletionException(new StreamSegmentSealedException(handle.getSegmentName(), exc));
```

Comment:
This situation arises only when two hosts are writing to the same file (in case of failover) and one of them sets the file readonly (by sealing the segment).